Add `contains` to `std/macrocache`

ire4ever1190 opened the pull request:
Implements `contains` for `CacheTable` mentioned in #19494

```nim
import std/[macrocache, macros]

let table = CacheTable"tableEx"

table["foo"] = newLit(42)
assert "foo" in table
```

`CacheTable` has it implemented similarly to `std/tables` where the `hasKey` proc is used to check if a key exists which is aliased by the `contains` proc.

ire4ever1190 committed:
Add test cases

ire4ever1190 committed:
Implement contains for CacheSeq

ire4ever1190 committed:
Implement contains for CacheTable

ire4ever1190 committed:
Fix implementation of hasKey

Araq commented on lib/core/macrocache.nim:
```diff
@@ -143,6 +143,19 @@ iterator items*(s: CacheSeq): NimNode =
 
   for i in 0 ..< len(s): yield s[i]
 
+proc contains*(s: CacheSeq, target: NimNode): bool {.inline.} =
```
Problematic because `NimNode` has reference semantics. Usually one would create a tree of identical shape though and then using `contains` cannot work.

ire4ever1190 committed:
Remove contains for CacheSeq

Fix runnable examples

I was accidently using --doccmd:skip so I didn't spot the failure locally

Araq commented on compiler/vm.nim:
```diff
@@ -2205,6 +2205,13 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
           stackTrace(c, tos, pc, "key does not exist: " & key)
       else:
         stackTrace(c, tos, pc, "key does not exist: " & destKey)
+    of opcNctContains:
```
Can this be implemented as a callback instead?

ringabout replied:
I think it is feasible since `PCtx` is accessible in function `proc registerAdditionalOps*(c: PCtx)`. We might extend it to similar cases since `opcNctLen`, etc. only needs to access `c.graph`. While we are at it, a benchmark will be appreciated.

ire4ever1190 replied:
Not too sure if I implemented it correctly but got it working as a callback

Using this quickly whipped up benchmark
```nim
import std/[times, macros, macrocache, math]

const
  t = CacheTable"l"
  n = 1_000

template timeIt(title: string, body: untyped) =
  var times: array[10, float]
  for i in 0..<10:
    let start = cpuTime()
    for i in 0..<n:
      body
    times[i] = cpuTime() - start
  echo title, ": ", sum(times) / len(times).float

proc main() =
  timeIt "Using magic":
    doAssert t.hasKeyMag("test")

  timeIt "Using VM hook":
    doAssert t.hasKey("test")

static:
  t["test"] = newLit 8
  main()
```
I get these numbers
```
Using magic: 0.0008407110000000717
Using VM hook: 0.001223563300000041
```

ringabout replied:
Nice!

ire4ever1190 committed:
Implement hasKey as a VM callback instead of magic

ringabout commented on lib/core/macrocache.nim:
```diff
@@ -181,6 +181,31 @@ proc `[]`*(t: CacheTable; key: string): NimNode {.magic: "NctGet".} =
       assert mcTable["toAdd"].kind == nnkStmtList
 
+proc hasKey*(t: CacheTable; key: string): bool =
+  ## Returns true if `key` is in the table `t`
```
```suggestion
  ## Returns true if `key` is in the table `t`.
```

ringabout commented on lib/core/macrocache.nim:
```diff
@@ -181,6 +181,31 @@ proc `[]`*(t: CacheTable; key: string): NimNode {.magic: "NctGet".} =
+      mcTable["foo"] = newEmptyNode()
+      # Will now be true since we inserted a value
+      assert mcTable.hasKey("foo")
+
```
You might give a comment or a discard statement telling it is implemented in the vmops.

ringabout commented on lib/core/macrocache.nim:
```diff
@@ -181,6 +181,31 @@ proc `[]`*(t: CacheTable; key: string): NimNode {.magic: "NctGet".} =
+      assert mcTable.hasKey("foo")
+
+proc contains*(t: CacheTable; key: string): bool =
+  ## Alias of [hasKey][hasKey(CacheTable, string)] for use with the `in` operator
```
```suggestion
  ## Alias of [hasKey][hasKey(CacheTable, string)] for use with the `in` operator.
```

ringabout commented on lib/core/macrocache.nim:
```diff
@@ -181,6 +181,31 @@ proc `[]`*(t: CacheTable; key: string): NimNode {.magic: "NctGet".} =
+      # Will now be true since we inserted a value
+      assert mcTable.hasKey("foo")
+
+proc contains*(t: CacheTable; key: string): bool =
```
```suggestion
proc contains*(t: CacheTable; key: string): bool {.inline.} =
```

ire4ever1190 committed:
Implement suggestions from PR

Co-Authored-By: ringabout <ringabout@users.noreply.github.com>

ringabout commented on lib/core/macrocache.nim:
```diff
@@ -181,6 +181,32 @@ proc `[]`*(t: CacheTable; key: string): NimNode {.magic: "NctGet".} =
+      mcTable["foo"] = newEmptyNode()
+      # Will now be true since we inserted a value
+      assert mcTable.hasKey("foo")
+  discard # Implemented in vmops
```
```suggestion
  discard "Implemented in vmops"
```

ringabout committed:
Update lib/core/macrocache.nim

ringabout reviewed (approved):
Excellent work!

ringabout committed:
Merge branch 'devel' into macrocache-contains

Araq commented:
Why is the CI red?

ire4ever1190 commented:
Red since datamancer has its own contains defined for `CacheTable`

Waiting on this PR to be merged https://github.com/SciNim/Datamancer/pull/49

Vindaar commented:
Just tagged a new version with the fix. CI should hopefully pass now.
